File: keyple-plugin/android-omapi/src/test/java/org/eclipse/keyple/plugin/android/omapi/AndroidOmapiReaderTest.java
/*
 * Copyright (c) 2018 Calypso Networks Association https://www.calypsonet-asso.org/
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse
 * Public License version 2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 */

package org.eclipse.keyple.plugin.android.omapi;


import static org.powermock.api.mockito.PowerMockito.when;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import org.eclipse.keyple.calypso.command.po.PoRevision;
import org.eclipse.keyple.calypso.command.po.builder.ReadRecordsCmdBuild;
import org.eclipse.keyple.seproxy.ApduRequest;
import org.eclipse.keyple.seproxy.ProxyReader;
import org.eclipse.keyple.seproxy.SeRequest;
import org.eclipse.keyple.seproxy.SeRequestSet;
import org.eclipse.keyple.seproxy.SeResponseSet;
import org.eclipse.keyple.seproxy.exception.KeypleBaseException;
import org.eclipse.keyple.seproxy.exception.KeypleReaderException;
import org.eclipse.keyple.seproxy.exception.NoStackTraceThrowable;
import org.eclipse.keyple.seproxy.protocol.ContactlessProtocols;
import org.eclipse.keyple.seproxy.protocol.ContactsProtocols;
import org.eclipse.keyple.util.ByteArrayUtils;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.powermock.modules.junit4.PowerMockRunner;
import org.simalliance.openmobileapi.Channel;
import org.simalliance.openmobileapi.Reader;
import org.simalliance.openmobileapi.Session;

@RunWith(PowerMockRunner.class)
public class AndroidOmapiReaderTest {

    final String PLUGIN_NAME = "AndroidOmapiPlugin";
    final String poAid = "A000000291A000000191";
    final String poAidResponse =
            "6F25840BA000000291A00000019102A516BF0C13C70800000000C0E11FA653070A3C230C1410019000";

    Reader omapiReader;
    ProxyReader proxyReader;

    // init before each test
    @Before
    public void SetUp() throws Exception {
        // default reader connected with secure element with poAid
        omapiReader = mockReader();

        // instanciate proxyReader with omapiReader
        proxyReader = new AndroidOmapiReader(PLUGIN_NAME, omapiReader, omapiReader.getName());
    }


    /*
     * TEST READERS
     */

    @Test
    public void getInstance() throws Exception {
        Assert.assertNotNull(proxyReader);
    }

    @Test
    public void getName() throws Exception {
        Assert.assertEquals(omapiReader.getName(), proxyReader.getName());
    }

    @Test
    public void isSEPresent() throws NoStackTraceThrowable {
        Assert.assertEquals(true, proxyReader.isSePresent());
    }

    @Test
    public void getParameters() throws NoStackTraceThrowable {
        Assert.assertNotNull(proxyReader.getParameters());
    }

    @Test
    public void setParameters() throws KeypleBaseException {
        Map<String, String> parameters = new HashMap<String, String>();
        parameters.put("key1", "value1");
        proxyReader.setParameters(parameters);
        Assert.assertTrue(proxyReader.getParameters().size() == 1);
        Assert.assertTrue(proxyReader.getParameters().get("key1").equals("value1"));
    }

    @Test
    public void setParameter() throws KeypleBaseException {
        proxyReader.setParameter("key2", "value2");
        Assert.assertTrue(proxyReader.getParameters().size() == 1);
        Assert.assertTrue(proxyReader.getParameters().get("key2").equals("value2"));
    }

    /*
     * TRANSMIT
     */

    @Test
    public void transmitHoplinkSuccessfull() throws KeypleBaseException {
        // default init

        // test
        SeResponseSet seResponse = proxyReader.transmit(getCalypsoRequestSample());

        // assert
        Assert.assertTrue(
                seResponse.getResponses().get(0).getApduResponses().get(0).isSuccessful());

    }

    @Test
    public void transmitNoAid() throws KeypleBaseException, IOException {

        // init
        omapiReader = mockReaderWithNoAid();
        proxyReader = new AndroidOmapiReader(PLUGIN_NAME, omapiReader, omapiReader.getName());

        // test
        SeResponseSet seResponse = proxyReader.transmit(getCalypsoRequestSample());

        // assert
        Assert.assertNull(seResponse.getResponses().get(0));

    }

    @Test
    public void transmitWrongProtocol() throws KeypleBaseException {
        // init
        String poAid = "A000000291A000000191";
        ReadRecordsCmdBuild poReadRecordCmd_T2Env = new ReadRecordsCmdBuild(PoRevision.REV3_1,
                (byte) 0x14, (byte) 0x01, true, (byte) 0x20, "Hoplink EF T2Environment");
        List<ApduRequest> poApduRequestList = Arrays.asList(poReadRecordCmd_T2Env.getApduRequest());

        // wrong protocol
        SeRequest seRequest =
                new SeRequest(new SeRequest.AidSelector(ByteArrayUtils.fromHex(poAid)),
                        poApduRequestList, false, ContactlessProtocols.PROTOCOL_MIFARE_UL);

        // test
        SeResponseSet seResponse = proxyReader.transmit(new SeRequestSet(seRequest));

        // assert
        Assert.assertNull(seResponse.getResponses().get(0));

    }

    @Test(expected = KeypleReaderException.class)
    public void transmitNotConnected() throws KeypleBaseException, IOException {

        // init
        Reader omapiReader = Mockito.mock(Reader.class);
        when(omapiReader.getName()).thenReturn("SIM1");
        when(omapiReader.isSecureElementPresent()).thenReturn(false);
        when(omapiReader.openSession()).thenThrow(new IOException());
        proxyReader = new AndroidOmapiReader(PLUGIN_NAME, omapiReader, omapiReader.getName());

        // test
        SeResponseSet seResponse = proxyReader.transmit(getCalypsoRequestSample());

        // expected = KeypleReaderException.class
    }



    /*
     * HELPERS
     */


    Reader mockReader() throws IOException {

        Reader omapiReader = Mockito.mock(Reader.class);
        Session session = Mockito.mock(Session.class);
        Channel channel = Mockito.mock(Channel.class);

        when(omapiReader.getName()).thenReturn("SIM1");
        when(omapiReader.isSecureElementPresent()).thenReturn(true);
        when(session.openLogicalChannel(ByteArrayUtils.fromHex(poAid))).thenReturn(channel);
        when(omapiReader.openSession()).thenReturn(session);
        when(session.getATR()).thenReturn(null);
        when(channel.getSelectResponse()).thenReturn(ByteArrayUtils.fromHex(poAidResponse));
        when(channel.getSession()).thenReturn(session);

        when(channel.transmit(ByteArrayUtils.fromHex("00B201A420"))).thenReturn(ByteArrayUtils
                .fromHex("00000000000000000000000000000000000000000000000000000000000000009000"));

        return omapiReader;

    }

    Reader mockReaderWithNoAid() throws KeypleBaseException, IOException {

        Reader omapiReader = Mockito.mock(Reader.class);
        Session session = Mockito.mock(Session.class);
        Channel channel = Mockito.mock(Channel.class);

        when(omapiReader.getName()).thenReturn("SIM1");
        when(omapiReader.isSecureElementPresent()).thenReturn(true);
        when(omapiReader.openSession()).thenReturn(session);
        when(session.openLogicalChannel(ByteArrayUtils.fromHex(poAid)))
                .thenThrow(new NoSuchElementException(""));

        return omapiReader;

    }

    SeRequestSet getCalypsoRequestSample() {
        String poAid = "A000000291A000000191";

        ReadRecordsCmdBuild poReadRecordCmd_T2Env = new ReadRecordsCmdBuild(PoRevision.REV3_1,
                (byte) 0x14, (byte) 0x01, true, (byte) 0x20, "Hoplink EF T2Environment");

        List<ApduRequest> poApduRequestList;

        poApduRequestList = Arrays.asList(poReadRecordCmd_T2Env.getApduRequest());

        SeRequest seRequest =
                new SeRequest(new SeRequest.AidSelector(ByteArrayUtils.fromHex(poAid)),
                        poApduRequestList, false, ContactsProtocols.PROTOCOL_ISO7816_3);

        return new SeRequestSet(seRequest);

    }


}

File: android/keyple-plugin/android-omapi/src/test/kotlin/org/eclipse/keyple/plugin/android/omapi/se/AndroidOmapiReaderTest.kt
/********************************************************************************
 * Copyright (c) 2020 Calypso Networks Association https://www.calypsonet-asso.org/
 *
 * See the NOTICE file(s) distributed with this work for additional information regarding copyright
 * ownership.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse
 * Public License 2.0 which is available at http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 ********************************************************************************/
package org.eclipse.keyple.plugin.android.omapi.se

import android.se.omapi.Channel
import android.se.omapi.Reader
import android.se.omapi.SEService
import android.se.omapi.Session
import io.mockk.every
import io.mockk.mockk
import org.eclipse.keyple.core.seproxy.plugin.reader.util.ContactsCardCommonProtocols
import org.eclipse.keyple.core.util.ByteArrayUtil
import org.eclipse.keyple.plugin.android.omapi.AbstractAndroidOmapiReaderTest
import org.eclipse.keyple.plugin.android.omapi.AndroidOmapiSupportedProtocols

internal class AndroidOmapiReaderTest : AbstractAndroidOmapiReaderTest<Reader, AndroidOmapiReader>() {

    override lateinit var nativeReader: Reader
    override lateinit var reader: AndroidOmapiReader

    override fun buildOmapiReaderImpl(nativeReader: Reader): AndroidOmapiReader {
        var androidOmapiReader = AndroidOmapiReader(nativeReader, PLUGIN_NAME, nativeReader.name)
        androidOmapiReader.activateProtocol(AndroidOmapiSupportedProtocols.ISO_7816_3.name, ContactsCardCommonProtocols.ISO_7816_3.name)
        return androidOmapiReader
    }

    override fun getNativeReaderName(): String {
        return reader.name
    }

    override fun mockReader(): Reader {
        val nativeReader = mockk<Reader>()
        val session = mockk<Session>()
        val seService = mockk<SEService>()
        val channel = mockk<Channel>()
        val version = "3.2"

        every { nativeReader.name } returns "SIM1"
        every { nativeReader.isSecureElementPresent } returns true
        every { nativeReader.seService } returns seService
        every { nativeReader.openSession() } returns session
        every { session.openLogicalChannel(ByteArrayUtil.fromHex(PO_AID), 0x00.toByte()) } returns channel
        every { session.isClosed } returns false
        every { seService.version } returns version
        every { seService.isConnected } returns true
        every { session.atr } returns ByteArrayUtil.fromHex("")
        every { session.closeChannels() } returns Unit
        every { channel.selectResponse } returns ByteArrayUtil.fromHex(PO_AID_RESPONSE)
        every { channel.session } returns session
        every { channel.session.close() } returns Unit
        every { channel.transmit(any()) } returns ByteArrayUtil.fromHex("00000000000000000000000000000000000000000000000000000000000000009000")

        return nativeReader
    }

    override fun mockReaderWithNoAid(): Reader {

        val nativeReader = mockk<Reader>()
        val session = mockk<Session>()
        val seService = mockk<SEService>()
        val version = "2.04"

        every { nativeReader.name } returns "SIM1"
        every { nativeReader.isSecureElementPresent } returns true
        every { session.isClosed } returns false
        every { session.atr } returns ByteArrayUtil.fromHex("")
        every { session.closeChannels() } returns Unit
        every { session.openLogicalChannel(ByteArrayUtil.fromHex(PO_AID), 0) } throws NoSuchElementException("")
        every { nativeReader.seService } returns seService
        every { nativeReader.openSession() } returns session
        every { seService.version } returns version
        return nativeReader
    }

    override fun mockReaderWithNullOnOpenLogicalChannel(): Reader {

        val nativeReader = mockk<Reader>()
        val session = mockk<Session>()
        val seService = mockk<SEService>()

        every { nativeReader.name } returns "SIM1"
        every { session.isClosed } returns false
        every { session.atr } returns ByteArrayUtil.fromHex("")
        every { session.closeChannels() } returns Unit
        every { session.openLogicalChannel(ByteArrayUtil.fromHex(PO_AID), 0) } returns null
        every { nativeReader.openSession() } returns session
        every { seService.isConnected } returns true
        return nativeReader
    }

    override fun mockReaderWithExceptionOnOpenLogicalChannel(throwable: Throwable): Reader {

        val nativeReader = mockk<Reader>()
        val session = mockk<Session>()
        val seService = mockk<SEService>()

        every { nativeReader.name } returns "SIM1"
        every { session.isClosed } returns false
        every { session.atr } returns ByteArrayUtil.fromHex("")
        every { session.closeChannels() } returns Unit
        every { session.openLogicalChannel(ByteArrayUtil.fromHex(PO_AID), 0) } throws throwable
        every { nativeReader.openSession() } returns session
        every { seService.isConnected } returns true
        return nativeReader
    }

    override fun mockReaderWithNullOnOpenBasicChannel(): Reader {

        val nativeReader = mockk<Reader>()
        val session = mockk<Session>()
        val seService = mockk<SEService>()

        every { nativeReader.name } returns "SIM1"
        every { session.isClosed } returns false
        every { session.atr } returns ByteArrayUtil.fromHex("")
        every { session.closeChannels() } returns Unit
        every { session.openBasicChannel(null) } returns null
        every { nativeReader.openSession() } returns session
        every { seService.isConnected } returns true
        return nativeReader
    }

    override fun mockReaderWithExceptionOnOpenBasicChannel(throwable: Throwable): Reader {

        val nativeReader = mockk<Reader>()
        val session = mockk<Session>()
        val seService = mockk<SEService>()

        every { nativeReader.name } returns "SIM1"
        every { session.isClosed } returns false
        every { session.atr } returns ByteArrayUtil.fromHex("")
        every { session.closeChannels() } returns Unit
        every { session.openBasicChannel(null) } throws throwable
        every { nativeReader.openSession() } returns session
        every { seService.isConnected } returns true
        return nativeReader
    }

    override fun mockReaderWithExceptionOnOpenSession(throwable: Throwable): Reader {
        val nativeReader = mockk<Reader>()
        every { nativeReader.name } returns "SIM1"
        every { nativeReader.isSecureElementPresent } returns true
        every { nativeReader.openSession() } throws throwable
        return nativeReader
    }

    override fun mockReaderWithExceptionOnCloseChannel(throwable: Throwable): Reader {

        val nativeReader = mockk<Reader>()
        val session = mockk<Session>()
        val seService = mockk<SEService>()
        val channel = mockk<Channel>()
        val version = "3.2"

        every { nativeReader.name } returns "SIM1"
        every { nativeReader.isSecureElementPresent } returns true
        every { nativeReader.seService } returns seService
        every { nativeReader.openSession() } returns session
        every { session.openLogicalChannel(ByteArrayUtil.fromHex(PO_AID), 0x00.toByte()) } returns channel
        every { session.isClosed } returns false
        every { seService.version } returns version
        every { seService.isConnected } returns true
        every { session.atr } returns ByteArrayUtil.fromHex("")
        every { session.closeChannels() } returns Unit
        every { channel.selectResponse } returns ByteArrayUtil.fromHex(PO_AID_RESPONSE)
        every { channel.session } returns session
        every { channel.session.close() } throws throwable
        every { channel.transmit(any()) } returns ByteArrayUtil.fromHex("00000000000000000000000000000000000000000000000000000000000000009000")

        return nativeReader
    }

    override fun mockReaderWithExceptionWhileTransmittingApdu(throwable: Throwable): Reader {
        val nativeReader = mockk<Reader>()
        val session = mockk<Session>()
        val seService = mockk<SEService>()
        val channel = mockk<Channel>()
        val version = "3.2"

        every { nativeReader.name } returns "SIM1"
        every { nativeReader.isSecureElementPresent } returns true
        every { nativeReader.seService } returns seService
        every { nativeReader.openSession() } returns session
        every { session.openLogicalChannel(any(), any()) } returns channel
        every { session.isClosed } returns false
        every { seService.version } returns version
        every { seService.isConnected } returns true
        every { session.atr } returns ByteArrayUtil.fromHex("")
        every { session.closeChannels() } returns Unit
        every { channel.selectResponse } returns ByteArrayUtil.fromHex(PO_AID_RESPONSE)
        every { channel.session } returns session
        every { channel.session.close() } returns Unit
        every { channel.transmit(any()) } throws throwable
        return nativeReader
    }
}
